KOGITO-2113 Simplify binary build deployment

Kevin-Mok opened the pull request:
See: https://issues.redhat.com/browse/KOGITO-2113

This PR aims to simplify the deployment of binary builds by removing the need to do an `oc start-build` command. 

## Recognizing Binary Builds
When passing a local `target` directory as the source (where `target` is the name of the last directory in the path), it realizes that the user is trying to do a binary build and automatically recognizes the runtime and whether or not the build needs to be native. It also now selectively picks the files it needs to upload based on the determined binary build type (see [s2i code](https://github.com/kiegroup/kogito-images/blob/c07b75691c7fe425acb68cf36a04e62d6c77f6a2/modules/kogito-s2i-core/added/s2i-core#L176-L195)).  This reduces the upload time of the binaries by ~66%.

With these improvements, I chose to force the user to enter a source argument so as to prevent user errors (e.g.  forgetting to set Spring Boot runtime or [native flag](https://issues.redhat.com/browse/KOGITO-3029)), improve the deployment time and reduce overall input.

I realize that some of these features may seem superfluous to the initial task, but I think I got kind of carried away trying to simplify the deployment process as much as possible. Also since I was already tweaking relevant parts of the code, I figured it'd be easier to include these changes as part of this PR.

## KOGITO-3029
[KOGITO-3029](https://issues.redhat.com/browse/KOGITO-3029) is resolved with this PR:

> If build has been declared as non native but native binary is given, then should warn or error

In this case, a [log message is given](https://github.com/Kevin-Mok/kogito-cloud-operator/blob/cc9fe82dc85926d402269ca1ae1762415cdd6fd8/cmd/kogito/command/converter/native_converter.go#L46), and the build type is set to native.

> If build has been declared as native but non native binary is given, then should warn or error

In this case, an [error is thrown](https://github.com/Kevin-Mok/kogito-cloud-operator/blob/cc9fe82dc85926d402269ca1ae1762415cdd6fd8/cmd/kogito/command/converter/native_converter.go#L44). 

## KOGITO-3106
[KOGITO-3106](https://issues.redhat.com/browse/KOGITO-3106) is also resolved in this PR by detecting the presence of the native binary (`*-runner`) in `target` and treating the build as a native Quarkus binary build. If the user wants a JVM Quarkus build, they can remove the native binary to do so.

## Testing 
I updated/added unit tests for changed/new functions as well manually testing all the different build types on OpenShift. I came across an issue while trying to do an s2i build, but this issue is present on `master` as well. I created [a Zulip stream](https://kie.zulipchat.com/#narrow/stream/232676-kogito/topic/s2i.20exception) for this issue.

## Requirements
Many thanks for submiting your Pull Request :heart:! 

Please make sure that your PR meets the following requirements:

- [x] You have read the [contributors' guide](https://github.com/kiegroup/kogito-cloud-operator/blob/master/README.md#contributing-to-the-kogito-operator)
- [x] Pull Request title is properly formatted: 
- [x] Pull Request contains link to the JIRA issue
- [x] Pull Request contains description of the issue
- [ ] Pull Request does not include fixes for issues other than the main ticket
- [x] Your feature/bug fix has a unit test that verifies it
- [x] You've tested the new feature/bug fix in an actual OpenShift cluster

Kevin-Mok commented:
It seems that the `pkg/apis/app/v1alpha1` package is failing its unit tests in the check, but I didn't change that part of the code. Also when running the tests locally, it seems to work fine:
```
~/go/src/github.com/kiegroup/kogito-cloud-operator on  KOGITO-2113
➜ go test github.com/kiegroup/kogito-cloud-operator/pkg/apis/app/v1alpha1
ok      github.com/kiegroup/kogito-cloud-operator/pkg/apis/app/v1alpha1 1.067s
```

ricardozanini commented:
@Kevin-Mok could you please verify the unit test?

Kevin-Mok commented:
I asked @Kaitou786 about the issue, and he also verified that the test was passing locally. I pushed the same commit to rerun the check, and it passes now. For reference, here was the failed test:
```
--- FAIL: TestSetProvisioning (0.00s)
    conditions_test.go:58: 
        	Error Trace:	conditions_test.go:58
        	Error:      	Should be true
        	Test:       	TestSetProvisioning
FAIL
coverage: 3.7% of statements
FAIL	github.com/kiegroup/kogito-cloud-operator/pkg/apis/app/v1alpha1	1.026s
```

ricardozanini commented on cmd/kogito/command/converter/binarybuildtype_converter_test.go:
```diff
@@ -0,0 +1,30 @@
+	"testing"
+)
+
+func Test_FromArgsToBinaryBuildType(t *testing.T) {
```
Don't need to change, but in the future you could consider table-driven tests: https://github.com/golang/go/wiki/TableDrivenTests

Kevin-Mok replied:
I changed it, but I'm not sure if I did it properly.

ricardozanini replied:
Better! Which IDE do you use? VSCode and Goland have a feature that you can generate an unit test for a given function. The base template is a table driven one. You can start from there.

Kevin-Mok replied:
I use Vim, but I found a [plugin](https://github.com/cweill/gotests) that has the same functionality. I updated the tests to use this templating.

ricardozanini commented on cmd/kogito/command/shared/mkdir.go:
```diff
@@ -0,0 +1,45 @@
+
+// TempDirWithFile creates a temporary directory dir with a
+// temporary file
+func TempDirWithFile(dir string, file string) string {
```
Is this function being used only on tests? If it's the case, can you please move to the `test` package where we have the shared test util functions?

ricardozanini commented on cmd/kogito/command/util/build_from_file.go:
```diff
@@ -21,69 +21,123 @@ import (
+var supportedJvmExtensions = []string{".jar", ".json"}
+var supportedNativeExtensions = []string{"-runner", ".json"}
+
+func zipFile(absoluteFilePath string, fileInfo os.FileInfo, resource string, binaryBuildType BinaryBuildType, tarWriter *tar.Writer) (string, error) {
```
I believe that those IO/build related functions could be refactored to their own package instead of `util`.

Kevin-Mok replied:
I moved this file and its test to an `iozip` package.

ricardozanini commented on pkg/client/openshift/buildconfig.go:
```diff
@@ -132,10 +132,13 @@ func (b *buildConfig) TriggerBuild(bc *buildv1.BuildConfig, triggeredBy string)
+
+	buildName := options.Name
+	if !binaryBuild {
+		buildName += "-builder"
```
`builder` is a constant for the KogitoBuild instance? I believe you can take that from the `build` package:
https://github.com/kiegroup/kogito-cloud-operator/blob/master/pkg/controller/kogitobuild/build/utils.go#L26

You might refactor this to be public.

Kevin-Mok replied:
I get a cyclical import when trying to import from the `build` package:
```
import cycle not allowed
package main
	imports github.com/kiegroup/kogito-cloud-operator/pkg/framework
	imports github.com/kiegroup/kogito-cloud-operator/pkg/client/openshift
	imports github.com/kiegroup/kogito-cloud-operator/pkg/controller/kogitobuild/build
	imports github.com/kiegroup/kogito-cloud-operator/pkg/client/openshift
import cycle not allowed
package main
	imports github.com/kiegroup/kogito-cloud-operator/pkg/framework
	imports github.com/kiegroup/kogito-cloud-operator/pkg/client/openshift
	imports github.com/kiegroup/kogito-cloud-operator/pkg/controller/kogitobuild/build
	imports github.com/kiegroup/kogito-cloud-operator/pkg/framework
```

ricardozanini replied:
You can export the constant to the `infrastructure` package and import it in both packages: https://docs.jboss.org/kogito/release/latest/html_single/#_kogito_operator_core_package_structure

Kevin-Mok replied:
There was another once I moved it to `infrastructure`...
```
import cycle not allowed
package main
	imports github.com/kiegroup/kogito-cloud-operator/pkg/infrastructure
	imports github.com/kiegroup/kogito-cloud-operator/pkg/framework
	imports github.com/kiegroup/kogito-cloud-operator/pkg/client/openshift
	imports github.com/kiegroup/kogito-cloud-operator/pkg/infrastructure
```

ricardozanini replied:
The constant is in the infra package and you can't use from the `cmd`? this is weird. Nevertheless, let it be then. We are passing through a major refactoring, I'll update later.

Kevin-Mok replied:
The file we're trying to import from is in `pkg`.

ricardozanini reviewed (commented):
Great job! Looks really neat, I have some questions, though. After you take care of those I'll come back and take another look. :)

ricardozanini commented on cmd/kogito/command/iozip/zip_files.go:
```diff
@@ -0,0 +1,169 @@
+var supportedJvmExtensions = []string{".jar", ".json"}
+var supportedNativeExtensions = []string{"-runner", ".json"}
+
+func zipFile(absoluteFilePath string, fileInfo os.FileInfo, resource string, binaryBuildType flag.BinaryBuildType, tarWriter *tar.Writer) (string, error) {
```
the package calls `iozip` and the function `zipFile`. Take care when naming packages to not have things like `iozip.Zip(...)`. In this case the function is private, but worth a heads up.

Kevin-Mok replied:
What would be wrong with `iozip.Zip()`?

ricardozanini replied:
I don't have the time to go through it now, but we have linters that check for package/function names. `zip.Zip` for instance. It's not a best practice. This was just a heads up since some linters can get you off guard.

ricardozanini commented on cmd/kogito/command/service/buildresource_service.go:
```diff
@@ -61,11 +61,14 @@ func GetResourceType(resource string) (ResourceType flag.ResourceType, err error
 		}
 		return "", fmt.Errorf("invalid resource %s", resource)
 	} else if fileInfo.Mode().IsDir() {
+		if fileInfo.Name() == "target" {
```
Consider adding `target` to a constant

ricardozanini reviewed (commented):
Nitpicks only :)

vaibhavjainwiz reviewed (approved):
LGTM

Kevin-Mok commented on cmd/kogito/command/iozip/zip_files.go:
```diff
@@ -0,0 +1,169 @@
+}
+
+// ProduceTGZfile produces a tgz file of the given directory files
+func ProduceTGZfile(resource string, binaryBuildType flag.BinaryBuildType) (io.Reader, error) {
```
Sure. I was just going with the legacy name, but `CompressAsTGZ` does sound better.

spolti reviewed (approved):
Nice work :)

ricardozanini commented:
/jenkins test

sutaakar commented on cmd/kogito/command/converter/runtimetype_converter.go:
```diff
@@ -17,9 +17,35 @@ package converter
+		}
+
+		for _, file := range files {
+			if strings.Contains(file.Name(), "-runner") {
```
```suggestion
			if strings.HasSuffix(file.Name(), "-runner") || strings.HasSuffix(file.Name(), "-runner.jar") {
```
This way we exclude files which contain `-runner` in the middle

Kevin-Mok replied:
Changed and updated the relevant Spring Boot test to check for a file with `-runner` in the middle as well.

sutaakar commented on cmd/kogito/command/flag/build_flag.go:
```diff
@@ -28,6 +28,8 @@ const (
 	// LocalDirectoryResource Build using directory on local system
 	LocalDirectoryResource ResourceType = "LocalDirectoryResource"
+	// LocalBinaryDirectoryResource Build using directory on local system
+	LocalBinaryDirectoryResource ResourceType = "LocalBinaryDirectoryResource"
```
What is the difference between `LocalBinaryDirectoryResource` and `LocalDirectoryResource` above? They have same comment.

sutaakar replied:
Does `LocalBinaryDirectoryResource` reference the folder contained built Kogito application

ricardozanini replied:
Good point, @sutaakar

sutaakar reviewed (approved):
The code looks good (good job), minor comments/clarifications added.
I will test the built CLI now.

ricardozanini commented:
/jenkins test

sutaakar commented:
BDD tests need to be updated. Current implementation creates "empty" build first and then upload binary using OC client. That doesn't work any more as binary upload is a mandatory part of CLI created binary build.

sutaakar commented:
I hit a problem - current implementation doesn't allow running KogitoRuntime using externally provided image (unless I missed something) - the current use case on Kubernetes.
The example command used to deploy KogitoRuntime:
`kogito deploy process-springboot-example --env JAVA_OPTIONS=-Xmx2G --image quay.io/ksuta/process-springboot-example-default-nightly:latest --replicas 1 --runtime springboot`

sutaakar reviewed (changes requested):
See comment above

Kevin-Mok commented:
It doesn't require a source argument in this case?

ricardozanini commented:
> It doesn't require a source argument in this case?

No because he is deploying with a built image. Won’t need a KogitoBuild for this case.

Kevin-Mok commented:
I've reverted the behavior.

sutaakar commented:
A nitpick:
In case binary build is uploaded (`kogito deploy process-springboot-example target ...`) the CLI shows this text in console:
`The requested file(s) was successfully uploaded to OpenShift, the build process-springboot-example-1 with this file(s) should now be running. To see the logs, run 'oc logs -f bc/process-springboot-example-builder -n ksuta-local-cucumber-f0fb'`
However this command is valid just for source file upload, in case of binary upload the correct command to get logs is 'oc logs -f bc/process-springboot-example -n ksuta-local-cucumber-f0fb'

ricardozanini commented:
/jenkins test

Kevin-Mok commented:
> A nitpick:
> In case binary build is uploaded (`kogito deploy process-springboot-example target ...`) the CLI shows this text in console:
> `The requested file(s) was successfully uploaded to OpenShift, the build process-springboot-example-1 with this file(s) should now be running. To see the logs, run 'oc logs -f bc/process-springboot-example-builder -n ksuta-local-cucumber-f0fb'`
> However this command is valid just for source file upload, in case of binary upload the correct command to get logs is 'oc logs -f bc/process-springboot-example -n ksuta-local-cucumber-f0fb'

Good catch. I'll fix this on Monday.

> /jenkins test

Seems the CI is down. I'll run the test after fixing the above.

sutaakar reviewed (approved):
Looks good to me.

sutaakar commented:
@Kevin-Mok Here is the BDD test coverage for the CLI binary upload - https://github.com/sutaakar/kogito-cloud-operator/commit/e5e8a93bf6f4015aea0757139a754729694c50ac
You can add it as part of this PR or can be merged separately

msmagnanijr commented:
/jenkins test

msmagnanijr commented:
/jenkins test

msmagnanijr commented:
/jenkins test

ricardozanini commented:
@Kevin-Mok please update the [RELEASE_NOTES.md](https://github.com/kiegroup/kogito-cloud-operator/blob/master/RELEASE_NOTES.md) file

msmagnanijr commented:
/jenkins test

msmagnanijr commented:
/jenkins test

msmagnanijr commented:
/jenkins test

msmagnanijr commented:
/jenkins test

msmagnanijr commented:
/jenkins test

msmagnanijr commented:
/jenkins test

Kevin-Mok commented:
> @Kevin-Mok Here is the BDD test coverage for the CLI binary upload - [sutaakar@e5e8a93](https://github.com/sutaakar/kogito-cloud-operator/commit/e5e8a93bf6f4015aea0757139a754729694c50ac)
> You can add it as part of this PR or can be merged separately

Added.

> @Kevin-Mok please update the [RELEASE_NOTES.md](https://github.com/kiegroup/kogito-cloud-operator/blob/master/RELEASE_NOTES.md?rgh-link-date=2020-09-22T12%3A27%3A51Z) file

Updated.

ricardozanini commented:
/jenkins test

msmagnanijr commented:
/jenkins test

ricardozanini commented:
@sutaakar looks like the tests are failing:
https://jenkins-kogito-tools.apps.kogito-cloud.automation.rhmw.io/blue/organizations/jenkins/kogito%2Fkogito-cloud-operator-PR-check/detail/kogito-cloud-operator-PR-check/46/tests

Infra related?

sutaakar commented:
@ricardozanini Those are https://issues.redhat.com/browse/KOGITO-3393 related, persistence is currently broken.
